### PR TITLE
`azurerm_mysql_flexible_server` - convert `public_network_access_enabled` from attribute to argument

### PR DESCRIPTION
**Pull request by wyattfry**
## Community Note

* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review

## PR Checklist

- [ ] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [ ] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [ ] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [ ] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [ ] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”

## Changes to existing Resource / Data Source

- [ ] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [ ] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [ ] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [ ] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.

## Testing 

- [ ] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

* `azurerm_resource` - support for the `thing1` property [GH-00000]

This is a (please select all that apply):

- [ ] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change

## Related Issue(s)
Fixes #0000

> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

### CODE REVIEW

File: internal/services/mysql/mysql_flexible_server_resource_test.go [hunk 1]
```diff
@@ -568,6 +568,35 @@ func TestAccMySqlFlexibleServer_updateToWriteOnlyPassword(t *testing.T) {
...
+func TestAccMySqlFlexibleServer_updatePublicNetworkAccess(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_mysql_flexible_server", "test")
+	r := MySqlFlexibleServerResource{}
+
+	data.ResourceTest(t, r, []acceptance.TestStep{
+		{
+			Config: r.publicNetworkAccess(data, "null"),
+			Check: acceptance.ComposeTestCheckFunc(
+				check.That(data.ResourceName).ExistsInAzure(r),
+			),
```
Comment (wyattfry) [inline internal/services/mysql/mysql_flexible_server_resource_test.go:580]: Are these `Check`s made unnecessary by the `ImportStep`s?

File: internal/services/mysql/mysql_flexible_server_resource_test.go [hunk 2]
```diff
@@ -1359,3 +1388,20 @@ resource "azurerm_mysql_flexible_server" "test" {
 }
 `, r.template(data), acceptance.WriteOnlyKeyVaultSecretTemplate(data, secret), data.RandomInteger, version)
 }
+
+func (r MySqlFlexibleServerResource) publicNetworkAccess(data acceptance.TestData, publicNetworkAccessEnabled string) string {
```
Comment (wyattfry) [inline internal/services/mysql/mysql_flexible_server_resource_test.go:1392]: NB: I made `publicNetworkAccessEnabled` a string so it could be `true`, `false` or `null`.

File: internal/services/mysql/mysql_flexible_server_resource_test.go (hunk 1 above)
Comment (mbfrahry) [reply internal/services/mysql/mysql_flexible_server_resource_test.go:580]: We have it here to confirm it [exists](https://github.com/hashicorp/terraform-provider-azurerm/pull/28890/files#diff-d24863cabe3a5a3e3e6acadc357ca64520901f1ffc25ce0bf48104065590a8b0R600) in Azure. Sometimes it doesn't even though it created because Azure is weird. You're probably right that we don't need it if we're checking import but it only costs an extra api call to doubly make sure it at least exists in Azure before we move on. Similar to how we had those old checks that we don't do anymore, we might move in that direction of removing this as well but for now, let's keep it consistent with how we do other tests

File: internal/services/mysql/mysql_flexible_server_resource_test.go [hunk 3]
```diff
@@ -568,6 +568,35 @@ func TestAccMySqlFlexibleServer_updateToWriteOnlyPassword(t *testing.T) {
 	})
 }
 
+func TestAccMySqlFlexibleServer_updatePublicNetworkAccess(t *testing.T) {
+	data := acceptance.BuildTestData(t, "azurerm_mysql_flexible_server", "test")
+	r := MySqlFlexibleServerResource{}
+
+	data.ResourceTest(t, r, []acceptance.TestStep{
+		{
+			Config: r.publicNetworkAccess(data, "null"),
```
Comment (mbfrahry) [inline internal/services/mysql/mysql_flexible_server_resource_test.go:577]: We normally don't test null but we might want to so I'm curious what are you hoping to see out of this test step?

File: internal/services/mysql/mysql_flexible_server_resource.go [hunk 4]
```diff
@@ -243,6 +243,12 @@ func resourceMysqlFlexibleServer() *pluginsdk.Resource {
 				ValidateFunc: privatezones.ValidatePrivateDnsZoneID,
 			},
 
+			"public_network_access_enabled": {
+				Type:     pluginsdk.TypeBool,
+				Optional: true,
+				Computed: true,
```
[the remark this answers is not in the document]
Comment (wyattfry) [reply internal/services/mysql/mysql_flexible_server_resource.go:249]: Yes, Azure defaults it to true

File: internal/services/mysql/mysql_flexible_server_resource_test.go (hunk 3 above)
Comment (wyattfry) [reply internal/services/mysql/mysql_flexible_server_resource_test.go:577]: Testing the default behavior. The Azure REST API has the public access enabled as optional, so it doesn't need to be sent, so this is just to ensure we don't have one of those Computed / Optional situations I like had before, where when not specified, the first plan after the initial create wants to change the setting to the zero value.

File: internal/services/mysql/mysql_flexible_server_resource_test.go (hunk 1 above)
Comment (wyattfry) [reply internal/services/mysql/mysql_flexible_server_resource_test.go:580]: Roger. I was considering removing them, but I looked through the code base and couldn't find any precedent of test steps without checks, so I left them in.

File: internal/services/mysql/mysql_flexible_server_resource.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (mbfrahry) [reply internal/services/mysql/mysql_flexible_server_resource.go:249]: Sweet! Then we can default it to True as well

File: internal/services/mysql/mysql_flexible_server_resource_test.go (hunk 3 above)
Comment (mbfrahry) [reply internal/services/mysql/mysql_flexible_server_resource_test.go:577]: Yea...so I think since Azure defaults public_network_access enabled to true, we can also default it to true and then don't have to test null

**Comment by wyattfry** [comment]
TODO: leave old `public_network_access_enabled` (bool) attribute, but begin its deprecation, and make a new argument called `public_network_access` (string) with possible values `Enabled`, `Disabled` and maybe a third thing.

File: internal/services/mysql/mysql_flexible_server_resource.go [hunk 5]
```diff
@@ -243,6 +243,12 @@ func resourceMysqlFlexibleServer() *pluginsdk.Resource {
 				ValidateFunc: privatezones.ValidatePrivateDnsZoneID,
 			},
 
+			"public_network_access_enabled": {
```
Comment (mbfrahry) [inline internal/services/mysql/mysql_flexible_server_resource.go:246]: One more note here, Microsoft is adding more options to `public_network_access_enabled` so a boolean won't work when that change gets added. We should rename this to `public_network_access` and have it be a string

Comment (wyattfry) [reply internal/services/mysql/mysql_flexible_server_resource.go:246]: Roger, thanks! Will do.

File: internal/services/mysql/mysql_flexible_server_resource.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (wyattfry) [reply internal/services/mysql/mysql_flexible_server_resource.go]: Thanks Neil! I will test it.

File: website/docs/r/mysql_flexible_server.html.markdown [hunk 6]
```diff
@@ -121,6 +121,8 @@ The following arguments are supported:
 
 ~> **NOTE:** The `private_dns_zone_id` is required when setting a `delegated_subnet_id`. The `azurerm_private_dns_zone` should end with suffix `.mysql.database.azure.com`.
 
+* `public_network_access` - (Optional) Whether public network access is allowed for the MySql Flexible Server. Possible values are `Enabled`, `Disabled`. Defaults to `Enabled`.
```
Comment (neil-yechenwei) [inline website/docs/r/mysql_flexible_server.html.markdown:124]: Seems `Defaults to `Enabled`.` needs to be removed.

File: internal/services/mysql/mysql_flexible_server_resource.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (wyattfry) [reply internal/services/mysql/mysql_flexible_server_resource.go]: Testing complete. If a subnet and private DNS zone are provided at time of creation, Azure sets `publicNetworkAccess` to `Disabled`, even if the request has it as `Enabled`.

File: website/docs/r/mysql_flexible_server.html.markdown (hunk 6 above)
Comment (wyattfry) [reply website/docs/r/mysql_flexible_server.html.markdown:124]: Good catch, thanks @neil-yechenwei . I will update the docs according to the changes.

Comment (wyattfry) [reply website/docs/r/mysql_flexible_server.html.markdown]: Done.

File: internal/services/mysql/mysql_flexible_server_resource_test.go (hunk 2 above)
Comment (wyattfry) [reply internal/services/mysql/mysql_flexible_server_resource_test.go]: This test now uses the respective Enum type.

File: website/docs/r/mysql_flexible_server.html.markdown (hunk 6 above)
Comment (wyattfry) [reply website/docs/r/mysql_flexible_server.html.markdown]: Technically, it does default to `Enabled` unless the server has VNet integration configured, in which case it defaults to `Disabled. I updated the documentation to explain this.

File: internal/services/mysql/mysql_flexible_server_resource.go [hunk 7]
```diff
@@ -237,6 +237,13 @@ func resourceMysqlFlexibleServer() *pluginsdk.Resource {
 				ValidateFunc: privatezones.ValidatePrivateDnsZoneID,
 			},
 
+			"public_network_access": {
+				Type:         pluginsdk.TypeString,
+				Optional:     true,
+				Default:      servers.EnableStatusEnumEnabled,
```
Comment (mbfrahry) [inline internal/services/mysql/mysql_flexible_server_resource.go:243]: Per @neil-yechenwei comment, if Azure doesn't default this consistently to Enabled, we shouldn't default this either as it'll break people's configuration for those who have it setup differently. I believe we can just do Optional/Computed. It shouldn't be a security risk since it normally defaults to True so people removing this from their config won't expect it to turn off

File: website/docs/r/mysql_flexible_server.html.markdown [hunk 8]
```diff
@@ -119,15 +119,19 @@ The following arguments are supported:
 
 * `private_dns_zone_id` - (Optional) The ID of the private DNS zone to create the MySQL Flexible Server. Changing this forces a new MySQL Flexible Server to be created.
 
-~> **NOTE:** The `private_dns_zone_id` is required when setting a `delegated_subnet_id`. The `azurerm_private_dns_zone` should end with suffix `.mysql.database.azure.com`.
+~> **Note:** The `private_dns_zone_id` is required when setting a `delegated_subnet_id`. The `azurerm_private_dns_zone` should end with suffix `.mysql.database.azure.com`.
+
+* `public_network_access` - (Optional) Whether approved public traffic is allowed through the firewall to this server. Defaults to `Enabled`.
```
Comment (mbfrahry) [inline website/docs/r/mysql_flexible_server.html.markdown:124]: From @neil-yechenwei comment, we can remove the Defaults to enabled as that will be a breaking change.

```suggestion
* `public_network_access` - (Optional) Whether approved public traffic is allowed through the firewall to this server.
```

File: internal/services/mysql/mysql_flexible_server_resource.go (hunk 7 above)
Comment (mbfrahry) [reply internal/services/mysql/mysql_flexible_server_resource.go:243]: Can you also leave an `// NOTE: O + C:` note for why it's Optional/Computed

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Hey @wyattfry, sorry for going back and forth on the Optional/Computed bit for this attribute. I've left some comments for why we should go back to that model

File: internal/services/mysql/mysql_flexible_server_resource.go (hunk 7 above)
Comment (wyattfry) [reply internal/services/mysql/mysql_flexible_server_resource.go]: Done!

File: internal/services/mysql/mysql_flexible_server_resource_test.go [hunk 9]
```diff
@@ -687,9 +687,9 @@ resource "azurerm_mysql_flexible_server" "test" {
     iops    = 360
   }
 
-  delegated_subnet_id = azurerm_subnet.test.id
-  private_dns_zone_id = azurerm_private_dns_zone.test.id
-  sku_name            = "GP_Standard_D2ds_v4"
+  delegated_subnet_id   = azurerm_subnet.test.id
+  private_dns_zone_id   = azurerm_private_dns_zone.test.id
+  sku_name              = "GP_Standard_D2ds_v4"
```
Comment (wyattfry) [inline internal/services/mysql/mysql_flexible_server_resource_test.go:692]:
```suggestion
  delegated_subnet_id = azurerm_subnet.test.id
  private_dns_zone_id = azurerm_private_dns_zone.test.id
  sku_name            = "GP_Standard_D2ds_v4"
```

File: internal/services/mysql/mysql_flexible_server_resource_test.go [hunk 10]
```diff
@@ -930,6 +930,7 @@ resource "azurerm_mysql_flexible_server" "test" {
   administrator_login    = "adminTerraform"
   administrator_password = "QAZwsx123"
   version                = "8.0.21"
+  public_network_access  = "Disabled"
```
Comment (wyattfry) [inline internal/services/mysql/mysql_flexible_server_resource_test.go:933]:
```suggestion
```

File: internal/services/mysql/mysql_flexible_server_resource.go [hunk 11]
```diff
@@ -335,6 +338,9 @@ func resourceMysqlFlexibleServer() *pluginsdk.Resource {
 		),
 	}
 
+	// TODO When using >= v0.71.1 of go-azure-helpers, remove next line and update the schema to use `OptionalComputed` zone validation
```
Comment (mbfrahry) [inline internal/services/mysql/mysql_flexible_server_resource.go:341]: Is this change needed to get `public_network_access` out the door? We should do separate PRs here and since it looks like we're about to update go-azure-helpers, can we instead skip this and make the change later?

**Review by mbfrahry — CHANGES_REQUESTED** [review]
Hey @wyattfry, left some more comments on how we can clean this up a little

File: internal/services/mysql/mysql_flexible_server_resource.go (hunk 11 above)
Comment (wyattfry) [reply internal/services/mysql/mysql_flexible_server_resource.go:341]: Not functionally needed, but it allows the `basic` and `publicNetworkAccess` acceptance tests to pass.

File: internal/services/mysql/mysql_flexible_server_resource.go [hunk 12]
```diff
@@ -237,10 +237,13 @@ func resourceMysqlFlexibleServer() *pluginsdk.Resource {
 				ValidateFunc: privatezones.ValidatePrivateDnsZoneID,
 			},
 
+			// This property is Optional & Computed because of the complex behavior in the Azure API:
+			// If values are provided for subnet and DNS, the API will set public network access to "Disabled",
+			// Otherwise, it will default to "Enabled".
```
[the remark this answers is not in the document]
Comment (wyattfry) [reply internal/services/mysql/mysql_flexible_server_resource.go]: Sorry about that, I didn't realize you wanted those exact words, I'll do that going forward. Don't you think we should also include what happens when using VNet integration? Rather than just saying it doesn't default to `Enabled`?

**Review by mbfrahry — COMMENTED** [review]
Thanks for the changes @wyattfry. I kicked off one of the tests that has the different default behavior and saw a failure. Could you take a look

```
=== NAME  TestAccMySqlFlexibleServer_complete
    testcase.go:173: Step 1/3 error: Error running apply: exit status 1

        Error: creating Flexible Server (Subscription: "1a6092a6-137e-4025-9a7c-ef77f76f2c02"
        Resource Group Name: "acctestRG-mysql-250326090502149179"
        Flexible Server Name: "acctest-fs-250326090502149179"): performing Create: unexpected status 400 (400 Bad Request) with error: InvalidParameterValue: Invalid value given for parameter publicNetworkAccess. Specify a valid parameter value.

          with azurerm_mysql_flexible_server.test,
          on terraform_plugin_test.tf line 83, in resource "azurerm_mysql_flexible_server" "test":
          83: resource "azurerm_mysql_flexible_server" "test" {
```

Looks like `1` is being passed in

```
"network": {
		         ...
			"publicNetworkAccess": "1"
},
```

**Comment by wyattfry** [comment]
> Thanks for the changes @wyattfry. I kicked off one of the tests that has the different default behavior and saw a failure. Could you take a look
> 
> ```
> === NAME  TestAccMySqlFlexibleServer_complete
>     testcase.go:173: Step 1/3 error: Error running apply: exit status 1
> 
>         Error: creating Flexible Server (Subscription: "1a6092a6-137e-4025-9a7c-ef77f76f2c02"
>         Resource Group Name: "acctestRG-mysql-250326090502149179"
>         Flexible Server Name: "acctest-fs-250326090502149179"): performing Create: unexpected status 400 (400 Bad Request) with error: InvalidParameterValue: Invalid value given for parameter publicNetworkAccess. Specify a valid parameter value.
> 
>           with azurerm_mysql_flexible_server.test,
>           on terraform_plugin_test.tf line 83, in resource "azurerm_mysql_flexible_server" "test":
>           83: resource "azurerm_mysql_flexible_server" "test" {
> ```
> 
> Looks like `1` is being passed in
> 
> ```
> "network": {
> 		         ...
> 			"publicNetworkAccess": "1"
> },
> ```

Thanks @mbfrahry , I will investigate.

**Comment by wyattfry** [comment]
> > Thanks for the changes @wyattfry. I kicked off one of the tests that has the different default behavior and saw a failure. Could you take a look
> > ```
> > === NAME  TestAccMySqlFlexibleServer_complete
> >     testcase.go:173: Step 1/3 error: Error running apply: exit status 1
> > 
> >         Error: creating Flexible Server (Subscription: "1a6092a6-137e-4025-9a7c-ef77f76f2c02"
> >         Resource Group Name: "acctestRG-mysql-250326090502149179"
> >         Flexible Server Name: "acctest-fs-250326090502149179"): performing Create: unexpected status 400 (400 Bad Request) with error: InvalidParameterValue: Invalid value given for parameter publicNetworkAccess. Specify a valid parameter value.
> > 
> >           with azurerm_mysql_flexible_server.test,
> >           on terraform_plugin_test.tf line 83, in resource "azurerm_mysql_flexible_server" "test":
> >           83: resource "azurerm_mysql_flexible_server" "test" {
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Looks like `1` is being passed in
> > ```
> > "network": {
> > 		         ...
> > 			"publicNetworkAccess": "1"
> > },
> > ```
> 
> Thanks @mbfrahry , I will investigate.

I re-ran the complete acceptance test with the latest commit and it passed.

**Review by catriona-m — DISMISSED** [review]
Thanks @wyattfry - I just noticed a couple of minor things but otherwise this looks good to me!

File: internal/services/mysql/mysql_flexible_server_resource.go [hunk 13]
```diff
@@ -39,16 +39,16 @@ const (
...
 		Read:   resourceMysqlFlexibleServerRead,
 		Update: resourceMysqlFlexibleServerUpdate,
 		Delete: resourceMysqlFlexibleServerDelete,
 
 		Timeouts: &pluginsdk.ResourceTimeout{
-			Create: pluginsdk.DefaultTimeout(1 * time.Hour),
+			Create: pluginsdk.DefaultTimeout(2 * time.Hour),
 			Read:   pluginsdk.DefaultTimeout(5 * time.Minute),
-			Update: pluginsdk.DefaultTimeout(1 * time.Hour),
+			Update: pluginsdk.DefaultTimeout(2 * time.Hour),
```
[the remark this answers is not in the document]
Comment (wyattfry) [reply internal/services/mysql/mysql_flexible_server_resource.go:51]: Thanks! Good catch.

File: internal/services/mysql/mysql_flexible_server_resource_test.go [hunk 14]
```diff
@@ -62,6 +61,7 @@ func TestAccMySqlFlexibleServer_requiresImport(t *testing.T) {
 
 func TestAccMySqlFlexibleServer_complete(t *testing.T) {
 	data := acceptance.BuildTestData(t, "azurerm_mysql_flexible_server", "test")
+	data.Locations.Primary = "eastus"
```
[the remark this answers is not in the document]
Comment (wyattfry) [reply internal/services/mysql/mysql_flexible_server_resource_test.go]: I restored it to the setup on `main`

File: internal/services/mysql/mysql_flexible_server_resource.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (wyattfry) [reply internal/services/mysql/mysql_flexible_server_resource.go:51]: Updated.

File: internal/services/mysql/mysql_flexible_server_resource.go [hunk 15]
```diff
@@ -539,9 +551,13 @@ func resourceMysqlFlexibleServerRead(d *pluginsdk.ResourceData, meta interface{}
 			d.Set("source_server_id", props.SourceServerResourceId)
 
 			if network := props.Network; network != nil {
-				d.Set("public_network_access_enabled", *network.PublicNetworkAccess == servers.EnableStatusEnumEnabled)
 				d.Set("delegated_subnet_id", network.DelegatedSubnetResourceId)
 				d.Set("private_dns_zone_id", network.PrivateDnsZoneResourceId)
+				d.Set("public_network_access", string(pointer.From(network.PublicNetworkAccess)))
+
+				if !features.FivePointOh() {
+					d.Set("public_network_access_enabled", *network.PublicNetworkAccess == servers.EnableStatusEnumEnabled)
```
Comment (sreallymatt) [inline internal/services/mysql/mysql_flexible_server_resource.go:559]: Since this is a pointer and we haven't nil checked it above, I think we should use `pointer.From` here as well.

```suggestion
					d.Set("public_network_access_enabled", pointer.From(network.PublicNetworkAccess) == servers.EnableStatusEnumEnabled)
```

File: website/docs/r/mysql_flexible_server.html.markdown [hunk 16]
```diff
@@ -119,15 +119,19 @@ The following arguments are supported:
 
 * `private_dns_zone_id` - (Optional) The ID of the private DNS zone to create the MySQL Flexible Server. Changing this forces a new MySQL Flexible Server to be created.
 
-~> **NOTE:** The `private_dns_zone_id` is required when setting a `delegated_subnet_id`. The `azurerm_private_dns_zone` should end with suffix `.mysql.database.azure.com`.
+~> **Note:** The `private_dns_zone_id` is required when setting a `delegated_subnet_id`. The `azurerm_private_dns_zone` should end with suffix `.mysql.database.azure.com`.
+
+* `public_network_access` - (Optional) Whether approved public traffic is allowed through the firewall to this server.
```
Comment (sreallymatt) [inline website/docs/r/mysql_flexible_server.html.markdown:124]: Since this property is an enum, we should list the possible values

```suggestion
* `public_network_access` - (Optional) Whether approved public traffic is allowed through the firewall to this server. Possible values are `Enabled` and `Disabled`.
```

**Review by sreallymatt — CHANGES_REQUESTED** [review]
Hey @wyattfry, I've left 2 additional suggestions inline

**Review by sreallymatt — APPROVED** [review]
Thanks @wyattfry, LGTM
